url4c41254592_topskychen_AuthTopk_tgz-pJ8-quickhull3d_QuickHull3DTestJ8/src/utility/geo/buildBtreeOfLines.java

```diff
@@ -0,0 +1,182 @@
+					float[] point = new float[2];
+					point[0] = Float.parseFloat(num[2]);
+					point[1] = Float.parseFloat(num[3]);
+					if(buildBtreeOfPoints.ISCAR){
+						point[0] = (int)(point[0] * buildBtreeOfPoints.SCALE);
+						point[1] = (int)(point[1] * buildBtreeOfPoints.SCALE);
+					}
+					points.add(point);
+					//if(DEBUG) System.err.println();
+				}
+			} catch (IOException e) {
+				// TODO Auto-generated catch block
+				e.printStackTrace();
+			}
+		} catch (FileNotFoundException e) {
+			// TODO Auto-generated catch block
+			e.printStackTrace();
+		}
+		delaunay = new Delaunay(points.toArray(new float[0][0]));
+		final long start = System.currentTimeMillis();
+		final int[] lock = new int[1];
+		lock[0] = 0;
+		final int totalNum = points.size();
+		for(int id = 0; id < ThreadNum; id ++){
+			threadStatus[id] = false;
+			final int tid  = id;
+			new Thread(new Runnable() {
+				@Override
+				public void run() {
+					// TODO Auto-generated method stub
+					int curid, threadId = tid;
+					while(true){
+						synchronized (lock) {
+							curid = lock[0];
+							lock[0]++;	
+						}
+						if(curid >= totalNum)break;
+						int[] pid = delaunay.getLinked(curid + 1);
+						for(int id : pid){
+							if(id == 0)continue;
+							int id1 = idmap.get(curid), id2 = idmap.get(id - 1);
+							if(id1 >= id2)continue;
+							Long lineId = DataOfLine.calcLineId(id1, id2);
+							Line nline = new Line(points.get(curid), points.get(id - 1));
+							DataOfLine nDataOfLine = new DataOfLine(lineId, nline);
+							nDataOfLine.signWithRSA(rsa);
+							btOfLine.put(lineId, nDataOfLine.writeToBytes());
+						}
+						if(curid % 200 == 0){
+							System.out.println("Commit Thread:\t" + threadId + "\tid:\t" + curid);
+							long end = System.currentTimeMillis();
+							long expectedTime = (end - start) / (curid + 1) * (totalNum - curid) / 1000;
+							System.out.println("Expected time remaining: " + expectedTime + " s = " + (expectedTime / 60) + " m = " + (expectedTime / 3600) + " h");
+							try {
+								recmanOfLine.commit();
+							} catch (IOException e) {
+								// TODO Auto-generated catch block
+								e.printStackTrace();
+							}
+						}
+						if(DEBUG){
+							//recmanOfLine.commit();
+							System.out.println("Thread :\t" + threadId + " id :\t" + curid + " has finished");
+						}
+					}
+					threadStatus[threadId] = true;
+				}
+			}).start();
+		}
+		while(true){
+			boolean found = false;
+			try {
+				Thread.sleep(100);
+			} catch (InterruptedException e) {
+				// TODO Auto-generated catch block
+				e.printStackTrace();
+			}
+			for(int i = 0; i < ThreadNum; i++){
+				if(threadStatus[i] == false){
+					found = true;
+				}
+			}
+			if(!found)break;
+		}
+		recmanOfLine.commit();
+	}
+	
+	public buildBtreeOfLines(boolean isLoad, String sourceFileName, String destFileName) throws IOException{
+		recmanOfLine = RecordManagerFactory.createRecordManager(destFileName);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code assigns a resource that might leak to a class field. This can cause a resource leak that slows down or crashes your system. For example, when buildBtreeOfLines.loadData throws an exception, and an if condition at line 142 evaluates to true, the *buildBtreeOfLines* object isn’t constructed so you can’t close this resource afterward.

**Fix**
Ensure that if an exception is thrown, the *recmanOfLine* resource is closed before the constructor is exited.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).